### PR TITLE
SRE-243 - Azure Multi-Region Workflows

### PR DESCRIPTION
**Pull request by jsclifford**
* Added Multi Region Workflows Templates for following workflows
    * im-deploy-az-app-manually
    * im-deploy-tf-manually
    * im-deploy-az-database-manually
    * im-run-tf-destroy

### CODE REVIEW

File: workflow-templates/im-deploy-tf-manual-apply.yml
Comment (mike-schenk) [inline workflow-templates/im-deploy-tf-manual-apply.yml:1]: The changes in this file should be in other PRs. One with the plan result summary, another one with the resource group lock stuff, and another with the app insights addition.

We should have a separate discussion about the use of resource group locks and whether it's a pattern the swat, pipeline (purple) and Azure (blue) teams will support, suggest or promote.

The app insights change is great. It probably doesn't even need to be commented-out. That will make it more like `im-deploy-az-app-manually`.

**Review by mike-schenk — COMMENTED** [review]
Part 1 of my review: move the changes that aren't for multi-region deployments to separate PRs.

File: workflow-templates/im-deploy-az-app-manually.yml
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-deploy-az-app-manually.yml:1]: I am curious why you want so many prs for the changes.  I was doing the original PRs to enhance them.  I will most likely remove those changes and not create PRs for the original workflows.  I will make that change and removal.

File: workflow-templates/im-deploy-tf-manual-apply.yml
Comment (jsclifford) [reply workflow-templates/im-deploy-tf-manual-apply.yml:1]: I have removed the changes so it just is the multi-region workflow in this pull request.

File: workflow-templates/im-build-tf-auto-plan-and-comment-on-prs.yml
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-build-tf-auto-plan-and-comment-on-prs.yml:1]: I have removed the changes from this pr.

File: workflow-templates/im-deploy-az-database.yml
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-deploy-az-database.yml:1]: I have removed the changes from this PR.

**Comment by jsclifford** [comment]
Not needed anymore with move resource in terraform

File: workflow-templates/im-deploy-az-app-manually.yml [hunk 1]
```diff
@@ -414,7 +509,7 @@ jobs:
             ]
 
       - name: Send Status to Deployment Notifications Channel for Prod Deploys
-        if: always() && env.ENVIRONMENT == 'prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
+        if: always() && inputs.environment-or-target == 'prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-app-manually.yml:512]:
```suggestion
        if: always() && needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT == 'Prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
```

File: workflow-templates/im-deploy-az-app-manually.yml [hunk 2]
```diff
@@ -371,7 +465,7 @@ jobs:
     needs: [set-vars, deploy-code]
     if: always()
     steps:
-      - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
+      - run: echo "The current environment is ${{ inputs.environment-or-target }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-app-manually.yml:468]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 3]
```diff
@@ -28,18 +28,18 @@ on:
   workflow_dispatch:
     inputs:
       tag:
-        description: The tag for the release that will be deployed.  For Production, only tags reachable by the default branch will be accepted.
-        required: true
-      environment:
-        description: The environment to deploy to
+        description: The tag for the release that will be deployed.  For Production, only tags reachable by the default branch will be accepted. Leave blank to use the workflow's branch.
+        required: false
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:32]: I think the `tag` input should continue to be required like in all the other deployment workflows, and the Verify Tag Exists step should still be used early in the workflow in the `set-vars` job.

File: workflow-templates/im-deploy-az-database.yml [hunk 4]
```diff
@@ -56,25 +56,36 @@ on:
   #   types: [<deployable_name>_deploy] # TODO: Replace <deployable_name>.  This will be used in the 'Deploy Multiple Items' workflow to target this deployment workflow.
 
 env:
-  ENVIRONMENT: ${{ github.event.inputs.environment }}
-  RELEASE_TAG: ${{ github.event.inputs.tag }} # This is the tag that we'll be deploying
+  ENVIRONMENT: ${{ inputs.environment-or-target }}
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:59]:
```suggestion
```

Shouldn't be used anymore.

File: workflow-templates/im-deploy-az-database.yml [hunk 5]
```diff
@@ -28,18 +28,18 @@ on:
...
         type: choice
-        options: # TODO: Update for the environments that are available
+        options:
           - dev
           - qa
-          - stage
           - demo
           - uat
+          - stage
           - prod
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:43]:
```suggestion
          - dev
          - qa
          - stage
          - demo
          - uat
          - prod
          - stage-secondary
          - prod-secondary
```

File: workflow-templates/im-deploy-az-database.yml [hunk 6]
```diff
@@ -56,25 +56,36 @@ on:
...
           echo $'
-          | Build Arguments | Value                    |
-          | ---             |  ---                     |
-          | Environment     | `${{ env.ENVIRONMENT }}` |
-          | Tag             | `${{ env.RELEASE_TAG }}` |' >> $GITHUB_STEP_SUMMARY
-                
+          | Deploy Arguments          | Value                    |
+          | ---                       |  ---                     |
+          | Environment               | `${{ env.ENVIRONMENT }}` |
+          | Release                   | `${{ env.RELEASE_TAG  }}`  |
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:86]:
```suggestion
          | Tag                       | `${{ env.RELEASE_TAG  }}`  |
```

File: workflow-templates/im-deploy-az-database.yml [hunk 7]
```diff
@@ -56,25 +56,36 @@ on:
...
+          | Workflow Branch/Tag       | `${{ github.ref_name }}` - SHA: `${{ github.sha }}` |' >> $GITHUB_STEP_SUMMARY
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:87]: I like this addition. It probably belongs in the other deployment templates that use `workflow_dispatch`.

Maybe change the label though:
```suggestion
          | Workflow Source           | `${{ github.ref_name }}` - SHA: `${{ github.sha }}` |' >> $GITHUB_STEP_SUMMARY
```

File: workflow-templates/im-deploy-az-database.yml [hunk 8]
```diff
@@ -85,26 +96,70 @@ jobs:
...
   # that environment, the workflow will continue without requiring anyone to approve the deployment.
   stakeholder-approval:
-    needs: [verify-tag-exists]
+    needs: [set-vars]
     runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
-    environment: '${{ github.event.inputs.environment }} Stakeholder Approval' # Use inputs context because env context is not available to environment:
+    environment: '${{ needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT }} Stakeholder Approval' # Use inputs context because env context is not available to environment:
     steps:
-      - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
+      - run: echo "The current environment is ${{ inputs.environment-or-target }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:150]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 9]
```diff
@@ -85,26 +96,70 @@ jobs:
...
   attestor-approval:
-    needs: [verify-tag-exists]
+    needs: [set-vars]
     runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
-    environment: '${{ github.event.inputs.environment }} Attestor Approval' # Use inputs context because env context is not available to environment:
+    environment: '${{ needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT }} Attestor Approval' # Use inputs context because env context is not available to environment:
     steps:
       - run: |
-          echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
+          echo "The current environment is ${{ inputs.environment-or-target }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:162]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 10]
```diff
@@ -116,16 +171,16 @@ jobs:
       # In this job, always checkout the default branch (not the branch that was provided as an input).  Also use
       # fetch-depth: 0 to retrieve the history and tags so we can check if a tag is reachable from the default branch.
       - name: Checkout Repository
-        if: env.ENVIRONMENT == 'prod'
+        if: inputs.environment-or-target == 'prod'
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:174]:
```suggestion
        if: needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT == 'Prod'
```

File: workflow-templates/im-deploy-az-database.yml [hunk 11]
```diff
@@ -116,16 +171,16 @@ jobs:
...
         with:
-          ref: 'main' # TODO: verify the name of your default branch
+          ref: 'main'  # TODO: verify the name of your default branch
           fetch-depth: 0
 
       - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
 
       - uses: im-open/is-tag-reachable-from-default-branch@v1.1
-        if: env.ENVIRONMENT == 'prod'
+        if: inputs.environment-or-target == 'prod'
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:183]:
```suggestion
        if: needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT == 'Prod'
```

File: workflow-templates/im-deploy-az-database.yml [hunk 12]
```diff
@@ -116,16 +171,16 @@ jobs:
...
       - name: Checkout Repository
-        if: env.ENVIRONMENT == 'prod'
+        if: inputs.environment-or-target == 'prod'
         uses: actions/checkout@v3
         with:
-          ref: 'main' # TODO: verify the name of your default branch
+          ref: 'main'  # TODO: verify the name of your default branch
           fetch-depth: 0
 
       - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:181]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 13]
```diff
@@ -135,11 +190,11 @@ jobs:
...
 
   deploy-az-db:
-    needs: [validate-tag-for-prod-deploys]
+    needs: [set-vars,validate-tag-for-prod-deploys]
     runs-on: [self-hosted, ubuntu-20.04]
-    environment: ${{ github.event.inputs.environment }} # Use inputs context because env context is not available to environment:
+    environment: ${{ needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT }}
     steps:
-      - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
+      - run: echo "The current environment is ${{ inputs.environment-or-target }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:197]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 14]
```diff
@@ -213,15 +269,26 @@ jobs:
...
           az cache purge
           az account clear
 
   update-deployment-board-and-send-teams-notification:
     runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
-    needs: [deploy-az-db]
+    needs: [set-vars,deploy-az-db]
     if: always()
     steps:
       - run: echo "The current environment is ${{ env.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:294]:
```suggestion
```

File: workflow-templates/im-deploy-az-database.yml [hunk 15]
```diff
@@ -267,7 +334,7 @@ jobs:
             ]
 
       - name: Send Status to Deployment Notifications Channel for Prod Deploys
-        if: always() && env.ENVIRONMENT == 'prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
+        if: always() && inputs.environment-or-target == 'prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
```
Comment (mike-schenk) [inline workflow-templates/im-deploy-az-database.yml:337]:
```suggestion
        if: always() && needs.set-vars.outputs.GITHUB_SECRETS_ENVIRONMENT == 'Prod' && steps.conclusion.outputs.workflow_conclusion == 'success'
```

**Review by mike-schenk — CHANGES_REQUESTED** [review]

File: workflow-templates/im-deploy-az-database.yml (hunk 3 above)
Comment (jsclifford) [reply workflow-templates/im-deploy-az-database.yml:32]: I do this in all my repos.  I use the workflow tag except for prod.  You have to type it in.  This lets people use the github ui to select the tag for all lower environments.

File: workflow-templates/im-deploy-az-app-manually.yml [hunk 16]
```diff
@@ -271,16 +328,45 @@ jobs:
         with:
           creds: ${{ secrets.AZURE_CREDENTIALS }} # This is an env-level secret
 
+      # TODO: Uncomment if you have azure locks in stage and prod
+      # - name: Delete RGRP Azure Locks
+      #   id: remove-locks
+      #   if: inputs.environment-or-target == 'prod' || inputs.environment-or-target == 'stage'
+      #   run: |
+      #     az group lock delete --name "${{ needs.set-vars.outputs.TARGET_RESOURCE_GROUP }}-delete-locks"  --resource-group "${{ needs.set-vars.outputs.TARGET_RESOURCE_GROUP }}"
+      #     sleep 5s
```
[the remark this answers is not in the document]
Comment (jsclifford) [reply workflow-templates/im-deploy-az-app-manually.yml:338]: Ok applied to all workflows I am editing.

**Review by mike-schenk — APPROVED** [review]

**Pull request merged**